File: app/src/main/java/com/savvasdalkitsis/gameframe/draw/view/LayersAdapter.java
package com.savvasdalkitsis.gameframe.draw.view;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import com.savvasdalkitsis.gameframe.model.Historical;
import com.savvasdalkitsis.gameframe.draw.model.Layer;
import com.savvasdalkitsis.gameframe.draw.model.LayerSettings;
import com.savvasdalkitsis.gameframe.draw.model.Model;

import java.util.Collections;
import java.util.List;

import rx.functions.Action1;

class LayersAdapter extends RecyclerView.Adapter<LayerViewHolder> {

    private Historical<Model> modelHistory;

    @Override
    public LayerViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new LayerViewHolder(parent);
    }

    @Override
    public void onBindViewHolder(LayerViewHolder holder, int position) {
        holder.clearListeners();
        holder.bind(layers().get(position));
        holder.setOnClickListener(v -> selectWithHistory(holder.getAdapterPosition()));
        holder.setOnLayerVisibilityChangedListener(visible ->
                modifyLayer(holder, layer -> layer.setVisible(visible)));
        holder.setOnItemDeletedListener(() -> removeLayer(holder));
        holder.setOnLayerDuplicatedListener(() -> duplicateLayer(holder));
        holder.setOnLayerSettingsClickedListener(() -> layerSettings(holder));
    }

    @Override
    public int getItemCount() {
        return layers().size();
    }

    private void modifyLayer(LayerViewHolder holder, Action1<Layer> layerModifier) {
        int position = holder.getAdapterPosition();
        progressTime();
        Layer layer = layers().get(position);
        layerModifier.call(layer);
        notifyItemChanged(position);
        notifyObservers();
    }

    private void selectWithHistory(int position) {
        progressTime();
        select(position);
    }

    private void select(int position) {
        int selectedItemPosition = selectedItemPosition();
        layers().get(selectedItemPosition).setSelected(false);
        layers().get(position).setSelected(true);
        notifyItemChanged(selectedItemPosition);
        notifyItemChanged(position);
        notifyObservers();
    }

    private void removeLayer(LayerViewHolder holder) {
        int position = holder.getAdapterPosition();
        progressTime();
        if (layers().get(position).isSelected()) {
            layers().get(position - 1).setSelected(true);
            notifyItemChanged(position - 1);
        }
        layers().remove(position);
        notifyItemRemoved(position);
        notifyObservers();
    }

    private void duplicateLayer(LayerViewHolder holder) {
        int position = holder.getAdapterPosition();
        Layer layer = layers().get(position);
        Layer newLayer = layer.replicateMoment();
        newLayer.getLayerSettings().setTitle(layer.getLayerSettings().getTitle() + " copy");
        addNewLayer(newLayer, position + 1);
    }

    void swapStarted() {
        modelHistory.progressTimeWithoutAnnouncing();
    }

    void swapLayers(RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
        int itemPosition = viewHolder.getAdapterPosition();
        int targetPosition = target.getAdapterPosition();
        Collections.swap(layers(), itemPosition, targetPosition);
        notifyItemMoved(itemPosition, targetPosition);
    }

    void swapLayersFinished() {
        modelHistory.collapsePresentWithPastIfTheSame();
        notifyObservers();
    }

    private void layerSettings(LayerViewHolder holder) {
        Context context = holder.itemView.getContext();
        LayerSettingsView.show(context, layers().get(holder.getAdapterPosition()), (ViewGroup) holder.itemView,
                layerSettings -> modifyLayer(holder, layer -> layer.setLayerSettings(layerSettings)));
    }

    void addNewLayer() {
        addNewLayer(Layer.create(LayerSettings.create().title("Layer " + layers().size())).build(), layers().size());
    }

    private void addNewLayer(Layer layer, int position) {
        progressTime();
        layers().add(position, layer);
        notifyItemInserted(position);
        select(position);
    }

    private List<Layer> layers() {
        return modelHistory.present().getLayers();
    }

    private void notifyObservers() {
        modelHistory.announcePresent();
    }

    private void progressTime() {
        modelHistory.progressTime();
    }

    void bind(Historical<Model> modelHistory) {
        this.modelHistory = modelHistory;
        modelHistory.observe().subscribe(model -> notifyDataSetChanged());
    }

    private int selectedItemPosition() {
        List<Layer> layers = layers();
        for (int i = 0; i < layers.size(); i++) {
            Layer layer = layers.get(i);
            if (layer.isSelected()) {
                return i;
            }
        }
        throw new IllegalStateException("Could not find selected layer");
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/feature/draw/view/LayersAdapter.kt
package com.savvasdalkitsis.gameframe.feature.draw.view

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.savvasdalkitsis.gameframe.feature.draw.model.Layer
import com.savvasdalkitsis.gameframe.feature.draw.model.LayerSettings
import com.savvasdalkitsis.gameframe.feature.draw.model.Model
import com.savvasdalkitsis.gameframe.infra.kotlin.findIndexOrThrow
import com.savvasdalkitsis.gameframe.feature.history.model.Historical
import java.util.*

internal class LayersAdapter : RecyclerView.Adapter<LayerViewHolder>() {

    private lateinit var modelHistory: Historical<Model>

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            LayerViewHolder(parent)

    override fun onBindViewHolder(holder: LayerViewHolder, position: Int) = with(holder) {
        clearListeners()
        bind(layers()[position])
        setOnClickListener { selectWithHistory(holder.adapterPosition) }
        setOnLayerVisibilityChangedListener { visible -> modifyLayer(holder, { it.isVisible = visible }) }
        setOnItemDeletedListener { removeLayer(holder) }
        setOnLayerDuplicatedListener { duplicateLayer(holder) }
        setOnLayerSettingsClickedListener { layerSettings(holder) }
    }

    override fun getItemCount() = layers().size

    private fun modifyLayer(holder: LayerViewHolder, layerModifier: (Layer) -> Unit) {
        val position = holder.adapterPosition
        progressTime()
        val layer = layers()[position]
        layerModifier(layer)
        notifyItemChanged(position)
        notifyObservers()
    }

    private fun selectWithHistory(position: Int) {
        progressTime()
        select(position)
    }

    private fun select(position: Int) {
        val selectedItemPosition = selectedItemPosition()
        layers()[selectedItemPosition].isSelected = false
        layers()[position].isSelected = true
        notifyItemChanged(selectedItemPosition)
        notifyItemChanged(position)
        notifyObservers()
    }

    private fun removeLayer(holder: LayerViewHolder) {
        val position = holder.adapterPosition
        progressTime()
        if (layers()[position].isSelected) {
            layers()[position - 1].isSelected = true
            notifyItemChanged(position - 1)
        }
        layers().removeAt(position)
        notifyItemRemoved(position)
        notifyObservers()
    }

    private fun duplicateLayer(holder: LayerViewHolder) {
        val position = holder.adapterPosition
        val layer = layers()[position]
        val newLayer = layer.replicateMoment()
        newLayer.layerSettings.title = layer.layerSettings.title + " copy"
        addNewLayer(newLayer, position + 1)
    }

    fun swapStarted() = modelHistory.progressTimeWithoutAnnouncing()

    fun swapLayers(viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder) {
        val itemPosition = viewHolder.adapterPosition
        val targetPosition = target.adapterPosition
        Collections.swap(layers(), itemPosition, targetPosition)
        notifyItemMoved(itemPosition, targetPosition)
    }

    fun swapLayersFinished() {
        modelHistory.collapsePresentWithPastIfTheSame()
        notifyObservers()
    }

    private fun layerSettings(holder: LayerViewHolder) {
        val context = holder.itemView.context
        LayerSettingsView.show(context, layers()[holder.adapterPosition], holder.itemView as ViewGroup,
                { layerSettings: LayerSettings -> modifyLayer(holder, { it.layerSettings = layerSettings }) } as LayerSettingsSetListener)
    }

    fun addNewLayer() = addNewLayer(Layer(layerSettings = LayerSettings(title = "Layer ${layers().size}")), layers().size)

    private fun addNewLayer(layer: Layer, position: Int) {
        progressTime()
        layers().add(position, layer)
        notifyItemInserted(position)
        select(position)
    }

    private fun layers() = modelHistory.present.layers

    private fun notifyObservers() = modelHistory.announcePresent()

    private fun progressTime() = modelHistory.progressTime()

    fun bind(modelHistory: Historical<Model>) {
        this.modelHistory = modelHistory
        modelHistory.observe().subscribe { notifyDataSetChanged() }
    }

    private fun selectedItemPosition() = layers().findIndexOrThrow(
            { it.isSelected },
            { IllegalStateException("Could not find selected layer") }
    )
}
